shared/actions/platform-specific/input-monitor.desktop.tsx

```diff
@@ -2,68 +2,118 @@ const timeToConsiderActiveForAwhile = 300000
 const timeToConsiderInactive = 60000
 
 type NotifyActiveFunction = (isActive: boolean) => void
-// 5 minutes after being active, consider us inactive after
-// an additional minute of no input
-// for the purpose of marking chat messages read
+type Reason = 'blur' | 'focus' | 'mouseKeyboard' | 'timeout'
+type State = 'appActive' | 'afterActiveCheck' | 'appInactive' | 'blurred' | 'focused'
+// State machine
+// appActive: User is active. Tell redux we're active. In 5 minutes go to 'afterActiveCheck'. If blur go 'appInactive' immediately
+// afterActiveCheck: User was 'appActive', in a window of 1 minute see if any keyboard / mouse happens. If so go to 'appInactive', else go appActive
+// appInactive: Wait for focus or keyboard/mouse, then go to 'appActive'. Tell redux we're inactive
+// blurred: App in background, wait for focus. Tell redux we're inactive
+// focused: App in foreground but no input yet, wait for keyboard/mouse
+
 class InputMonitor {
-  active = true
-  notifyActive: NotifyActiveFunction
-  activeTimeoutID?: number
-  inactiveTimeoutID?: number
+  notifyActive?: NotifyActiveFunction
+  private state: State
+  private timeoutID?: NodeJS.Timeout
 
-  constructor(notifyActive: NotifyActiveFunction) {
-    this.notifyActive = notifyActive
-    // go into listening mode again
-    window.addEventListener('focus', this.goListening)
-    window.addEventListener('blur', this.goInactive)
-    this.goListening()
+  constructor() {
+    window.addEventListener('focus', this.onFocus)
+    window.addEventListener('blur', this.onBlur)
+    this.state = 'appInactive'
+    this.transition('focus')
```

Comment:
Could it race? If user opens Keybase and make it background immediately, it seems it should not transition to focus here.

Reply:
i think its fine to start in this state. we have the mouse/keyboard listeners and when it detects something it goes into active